micropeak/micropeak.1
.\"
.\" Copyright © 2013 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH MICROPEAK 1 "micropeak" ""
.SH NAME
micropeak \- MicroPeak logging altimeter download and analysis
.SH SYNOPSIS
.B "micropeak"
.SH DESCRIPTION
.I micropeak
connects to a MicroPeak USB adapter for MicroPeak data download.
It provides a menu-oriented
user interface to download and analyze data logged on a MicroPeak
altimeter.
.SH USAGE
When connected to a MicroPeak USB adapter,
.I micropeak 
can download data from a MicroPeak device, save it to a file and graph
data from saved files.
.P
A number of other menu options exist, including the ability to export flight
data in different formats.
.SH FILES
All data log files are recorded into a user-specified directory
(default ~/AltusMetrum). Files are named using the current date and a
unique flight number with a '.mpd' extension.
.SH AUTHOR
Keith Packard


ChangeLog
commit 701c26ed85c28ac59e338975f2a6ba6bd25f6493
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 19 00:16:16 2013 -0800

    altosdroid: bump versionName to 1.3 and versionCode to 4
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9f95ffbad918a73cfd5460d6ce037d680465c35d
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 19 00:12:21 2013 -0800

    altosui: When device has no valid romconfig, set RF cal to 0
    
    This is intended to signal to the user that no valid value was found
    and that they'd best pick something sensible.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a04c1dd5df76c9127615bc797a9d9f764eec1234
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 19 00:08:50 2013 -0800

    altos/lpc: Stop sending SETUP IN when the requested size is reached
    
    The host won't keep asking for SETUP IN packets once it has received
    the amount of data requested, so check to see if we've sent that much
    and flip back to IDLE state if so.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1ab12861c3e70d7c22b27d988546a925616a0adc
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 23:27:34 2013 -0800

    altos/lpc: Reset less of the device on USB reset.
    
    This leaves most of the device configured across USB reset, which
    appears to help when sending a IN reply to the first SETUP packet;
    without this change, the IN reply would always get a length of 0,
    which is fine for SET_ADDRESS, but not for GET_DESCRIPTOR_DEVICE,
    which OS X appears to send before setting the address (go figure).
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 3b13cc2ca035b13582cd2e59ba7286f872f43c6e
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 22:00:13 2013 -0800

    altoslib: Remove some old debug printfs for self flashing
    
    These aren't necessary anymore and just slow down flashing boards.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 408b0dea338147382e94717dab85b4a204e7bdf5
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 21:08:33 2013 -0800

    micropeak: Add micropeak man page
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2b2ba87d5f68b9e052dddd49d69341f36d777122
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 21:02:15 2013 -0800

    ao-tools: Add man pages for ao-flash utilities
    
    These aren't very wordy, but these tools are pretty simple scripts.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 39cb8c2896317b7538353be979ac99baffc14489
Merge: 2a6016c ee42796
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 21:53:52 2013 -0700

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 2a6016cfabc8cd56f5219871e3b3df316a639289
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 21:53:36 2013 -0700

    update Debian standards version we claim compliance with

commit ee4279613b4757453d0d8f8afc06037c61eeb520
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 20:32:05 2013 -0800

    altos: Try IMU self-test 10 times before giving up
    
    This should keep the device from failing to boot unless the IMU is
    actually broken. Oh, and if self test does fail, this places the
    flight computer in 'Invalid' state rather than panic.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1bf84ec28a41f7bd1b11ba45b4639856266227bc
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 20:30:58 2013 -0800

    doc: Add tables describing AltOS beeps and flashes
    
    Provide a convenient place to reference when listening to the device.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0673344289772ed89483948184d6608c272c7c26
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 18:20:55 2013 -0800

    altos/stm: Semantic error in STM usb disable caused it to not work
    
    The USB enable register wasn't actually getting rewritten with the
    enable bit turned off, so the USB device was still powered on in flight.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 122f491e459b6ff417932370b3f1aa2091c71aca
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 18:30:54 2013 -0700

    update release docs to include option for submodules

commit d9982c257463f23be940eea66bd4dc3aadff0043
Merge: 1b97ed2 b63fc05
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 18:25:35 2013 -0700

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 1b97ed2b64bcbcd969124964f1e49837899f1c70
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 18:25:03 2013 -0700

    we're using packaged and local-to-our-tree ARM toolchain now

commit b63fc05481bf6d57e6385704ce53c1c19afa9c2e
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:34:31 2013 -0800

    doc: typo in micropeak doc hole->hold

commit 6827961c002757f8e74de44f6eb9c9029d099ebc
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:25:41 2013 -0800

    doc: Update micropeak quick start guide to note new boost detect
    
    Now waits for one minute and 30m of altitude change to avoid false detections.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c0966cd40f05f3a65b0c977b4b92586a58192f4b
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:22:51 2013 -0800

    micropeak: Compile for java 6
    
    Don't a require later version as not all target OSes support it
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit eea036650e62bc0f8652155974b512686754fd13
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:08:41 2013 -0800

    Move pdclib build results to pdclib-root
    
    This makes pdclib easier to manage as a submodule
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c1bfe09b6d3eb28d0c7cfe07a248843cf81bcd25
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 13:36:04 2013 -0800


micropeak/MicroPeak.java
/*
 * Copyright © 2012 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

package org.altusmetrum.micropeak;

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.io.*;
import java.util.concurrent.*;
import java.util.*;
import org.altusmetrum.altoslib_3.*;
import org.altusmetrum.altosuilib_1.*;

public class MicroPeak extends MicroFrame implements ActionListener, ItemListener {

	File		filename;
	MicroGraph	graph;
	AltosUIEnable	enable;
	MicroStatsTable	statsTable;
	MicroRaw	raw;
	MicroData	data;
	MicroStats	stats;
	Container	container;
	JTabbedPane	pane;
	static int	number_of_windows;

	MicroPeak SetData(MicroData data) {
		MicroPeak	mp = this;
		if (this.data != null) {
			mp = new MicroPeak();
			return mp.SetData(data);
		}
		this.data = data;
		stats = new MicroStats(data);
		graph.setDataSet(data);
		statsTable.setStats(stats);
		raw.setData(data);
		setTitle(data.name);
		return this;
	}

	void SetName(String name) {
		graph.setName(name);
		setTitle(name);
	}

	private static MicroData ReadFile(File filename) throws IOException, FileNotFoundException {
		MicroData	data = null;
		FileInputStream	fis = new FileInputStream(filename);
		try {
			data = new MicroData((InputStream) fis, filename.getName());
			AltosUIPreferences.set_last_logdir(filename);
		} catch (InterruptedException ie) {
			data = null;
		} finally {
			fis.close();
		}
		return data;
	}

	private void OpenFile(File filename) {
		try {
			SetData(ReadFile(filename));
		} catch (FileNotFoundException fne) {
			JOptionPane.showMessageDialog(this,
						      fne.getMessage(),
						      "Cannot open file",
						      JOptionPane.ERROR_MESSAGE);
		} catch (IOException ioe) {
			JOptionPane.showMessageDialog(this,
						      ioe.getMessage(),
						      "File Read Error",
						      JOptionPane.ERROR_MESSAGE);
		}
	}

	private void SelectFile() {
		MicroFileChooser	chooser = new MicroFileChooser(this);
		File			file = chooser.runDialog();

		if (file != null)
			OpenFile(file);
	}

	private void Preferences() {
		new AltosUIConfigure(this);
	}

	private void DownloadData() {
		AltosDevice	device = MicroDeviceDialog.show(this);
		
		if (device != null)
			new MicroDownload(this, device);
	}

	private void no_data() {
			JOptionPane.showMessageDialog(this,
						      "No data available",
						      "No data",
						      JOptionPane.INFORMATION_MESSAGE);
	}

	private void Save() {
		if (data == null) {
			no_data();
			return;
		}
		MicroSave	save = new MicroSave (this, data);
		if (save.runDialog())
			SetName(data.name);
	}
	
	private void Export() {
		if (data == null) {
			no_data();
			return;
		}
		MicroExport	export = new MicroExport (this, data);
		export.runDialog();
	}

	private static void CommandGraph(File file) {
		MicroPeak m = new MicroPeak();
		m.OpenFile(file);
	}

	private static void CommandExport(File file) {
		try {
			MicroData d = ReadFile(file);
			if (d != null) {
				File	csv = new File(AltosLib.replace_extension(file.getPath(), ".csv"));
				try {
					System.out.printf ("Export \"%s\" to \"%s\"\n", file.getPath(), csv.getPath());
					MicroExport.export(csv, d);
				} catch (FileNotFoundException fe) {
					System.err.printf("Cannot create file \"%s\" (%s)\n", csv.getName(), fe.getMessage());
				} catch (IOException ie) {
					System.err.printf("Cannot write file \"%s\" (%s)\n", csv.getName(), ie.getMessage());
				}
			}
		} catch (IOException ie) {
			System.err.printf("Cannot read file \"%s\" (%s)\n", file.getName(), ie.getMessage());
		}
	}

	private void Close() {
		setVisible(false);
		dispose();
		--number_of_windows;
		if (number_of_windows == 0)
			System.exit(0);
	}

	public void actionPerformed(ActionEvent ev) {
		if ("Exit".equals(ev.getActionCommand()))
			System.exit(0);
		else if ("Close".equals(ev.getActionCommand()))
			Close();
		else if ("Open".equals(ev.getActionCommand()))
			SelectFile();
		else if ("Download".equals(ev.getActionCommand()))
			DownloadData();
		else if ("Export".equals(ev.getActionCommand()))
			Export();
		else if ("Preferences".equals(ev.getActionCommand()))
			Preferences();
		else if ("Save a Copy".equals(ev.getActionCommand()))
			Save();
	}

	public void itemStateChanged(ItemEvent e) {
	}

	public MicroPeak() {

		++number_of_windows;

		AltosUIPreferences.set_component(this);

		container = getContentPane();
		pane = new JTabbedPane();

		setTitle("MicroPeak");

		JMenuBar menuBar = new JMenuBar();
		setJMenuBar(menuBar);

		JMenu fileMenu = new JMenu("File");
		menuBar.add(fileMenu);

		JMenuItem openAction = new JMenuItem("Open");
		fileMenu.add(openAction);
		openAction.addActionListener(this);

		JMenuItem downloadAction = new JMenuItem("Download");
		fileMenu.add(downloadAction);
		downloadAction.addActionListener(this);

		JMenuItem saveAction = new JMenuItem("Save a Copy");
		fileMenu.add(saveAction);
		saveAction.addActionListener(this);

		JMenuItem exportAction = new JMenuItem("Export");
		fileMenu.add(exportAction);
		exportAction.addActionListener(this);

		JMenuItem preferencesAction = new JMenuItem("Preferences");
		fileMenu.add(preferencesAction);
		preferencesAction.addActionListener(this);

		JMenuItem closeAction = new JMenuItem("Close");
		fileMenu.add(closeAction);
		closeAction.addActionListener(this);

		JMenuItem exitAction = new JMenuItem("Exit");
		fileMenu.add(exitAction);
		exitAction.addActionListener(this);

		JButton downloadButton = new JButton ("Download");
		downloadButton.addActionListener(this);
		menuBar.add(downloadButton);

		setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
		addWindowListener(new WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent e) {
				statsTable.tell_closing();
				Close();
			}
		});

		enable = new AltosUIEnable();
		graph = new MicroGraph(enable);
		statsTable = new MicroStatsTable();
		raw = new MicroRaw();
		pane.add(graph.panel, "Graph");
		pane.add(enable, "Configure Graph");
		pane.add(statsTable, "Statistics");
		JScrollPane scroll = new JScrollPane(raw);
		pane.add(scroll, "Raw Data");
		pane.doLayout();
		pane.validate();
		container.add(pane);
		container.doLayout();
		container.validate();
		doLayout();
		validate();
		Insets i = getInsets();
		Dimension ps = pane.getPreferredSize();
		ps.width += i.left + i.right;
		ps.height += i.top + i.bottom;
//		setPreferredSize(ps);
		setSize(ps);
		setVisible(true);
	}

	public static void help(int code) {
		System.out.printf("Usage: micropeak [OPTION] ... [FILE]...\n");
		System.out.printf("  Options:\n");
		System.out.printf("    --csv\tgenerate comma separated output for spreadsheets, etc\n");
		System.out.printf("    --graph\tgraph a flight\n");
		System.exit(code);
	}

	public static void main(final String[] args) {
		boolean	opened = false;
		boolean graphing = true;

		try {
			UIManager.setLookAndFeel(AltosUIPreferences.look_and_feel());
		} catch (Exception e) {
		}

		for (int i = 0; i < args.length; i++) {
			if (args[i].equals("--help"))
				help(0);
			else if (args[i].equals("--export"))
				graphing = false;
			else if (args[i].equals("--graph"))
				graphing = true;
			else if (args[i].startsWith("--"))
				help(1);
			else {
				File	file = new File(args[i]);
				try {
					if (graphing)
						CommandGraph(file);
					else
						CommandExport(file);
					opened = true;
				} catch (Exception e) {
					System.err.printf("Error processing \"%s\": %s\n",
							  file.getName(), e.getMessage());
				}
			}
		}
		if (!opened)
			new MicroPeak();
	}
}
